manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+	//
+	// NOTE(dperny): be careful with this! if this loop doesn't run (because
+	// there were no items in ipamConfigs) then this will fail silently!
+	for _, specConfig := range ipamConfigs {
+		config := specConfig.Copy()
+		// the last parameter of this is "v6 bool", but we don't support ipv6
+		// so we just pass "false"
+		poolID, poolIP, meta, err := ipam.RequestPool(addressSpace, config.Subnet, config.Range, ipamOpts, false)
+		if err != nil {
+			return ErrFailedPoolRequest{config.Subnet, config.Range, err}
+		}
+		local.pools[poolIP.String()] = poolID
+		// The IPAM contract allows the IPAM driver to autonomously provide a
+		// network gateway in response to the pool request.  But if the network
+		// spec contains a gateway, we will allocate it irrespective of whether
+		// the ipam driver returned one already.  If none of the above is true,
+		// we need to allocate one now, and let the driver know this request is
+		// for the network gateway.
+		var (
+			gwIP *net.IPNet
+			ip   net.IP
+		)
+
+		if gws, ok := meta[netlabel.Gateway]; ok {
+			if ip, gwIP, err = net.ParseCIDR(gws); err != nil {
+				return ErrBustedIPAM{
+					ipamName,
+					n.ID,
+					fmt.Errorf(
+						"can't parse gateway address (%v) returned by the ipam driver: %v",
+						gws, err,
+					),
+				}
+			}
+			gwIP.IP = ip
+		}
+		// add the option indicating that we're gonna request a gateway, and
+		// remove it before we exit this function
+		ipamOpts[ipamapi.RequestAddressType] = netlabel.Gateway
+		defer delete(ipamOpts, ipamapi.RequestAddressType)
+		if config.Gateway != "" || gwIP == nil {
+			gwIP, _, err = ipam.RequestAddress(poolID, net.ParseIP(config.Gateway), ipamOpts)
+			if err != nil {
+				return ErrFailedAddressRequest{config.Gateway, err}
+			}
+		}
+		if config.Subnet == "" {
+			config.Subnet = poolIP.String()
+		}
+		if config.Gateway == "" {
+			config.Gateway = gwIP.IP.String()
+		}
+		finalConfigs = append(finalConfigs, config)
+	}
+
+	// now that everythign has succeeded, add the fields to the network object
+	n.IPAM = &api.IPAMOptions{
+		Driver: &api.Driver{
+			Name:    ipamName,
+			Options: ipamOpts,
+		},
+	}
+	n.IPAM.Configs = finalConfigs
+
+	// finally, add this network to the set of allocated networks.
+	a.networks[n.ID] = local
+	return nil
+}
+
+// DeallocateNetwork takes a network that has been allocated, and releases all
+// of the IPAM resource associated with it.
+func (a *allocator) DeallocateNetwork(network *api.Network) {
+	local, ok := a.networks[network.ID]
+	if !ok {
+		// if the network was never allocated, nothing to do
+		return
+	}
+
+	// we know, because we allocated this network, that its IPAM field is fully
+	// allocated, and we can use it without nil checking
+	ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+
+	for _, config := range local.nw.IPAM.Configs {
+		// these things can return errors, but we literally can't do anything
+		// about it if they do, so just ignore it.
```

Comment:
It probably would be useful... but I'm honestly not sure what to do about it. I think that I need to modify this code so that errors on deallocation are kicked all the way up to the caller, who can log them. This avoids importing the logging library.